### DISCUSSION TITLE
GLSP editor on reload

### DISCUSSION THREAD

**Discussion by LokanandaHJ** [eclipse-glsp/glsp#940] [General]
![image](https://user-images.githubusercontent.com/97349849/222329438-cf14df37-c398-43c8-9fac-a95c56c72f50.png)

the markers that are present on the node will be removed in the GLSP whenever we reload.

![image](https://user-images.githubusercontent.com/97349849/222331251-ecbf8cb6-c902-4119-8b47-5a9fca8972d6.png)

eventhough the problems are there in problem view.
![image](https://user-images.githubusercontent.com/97349849/222333558-83c78357-ea2d-4fcb-abe0-79cd69b9ab9c.png)

node is missing property called cssClasses

before reload- 
![image](https://user-images.githubusercontent.com/97349849/222333124-3618d5a3-e386-452d-8a34-5f08e113d2f4.png)

after reload-
![image](https://user-images.githubusercontent.com/97349849/222333227-f2ec8ee4-561b-46b6-81ad-71e9818e81e6.png)

what might be the reason

**Comment by planger** [comment]
Hi @LokanandaHJ,

yes this is sort of inconsistent: the problem entries in the Theia problem view remain, while the markers in the diagram editors are not re-added after close and re-open.

Some background:
By default, the model is only validated when the client [requests markers](https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md#2101-requestmarkersaction) explicitly. If the client receives markers from the server, it'll add them to the diagram.
In the Theia integration, the [`TheiaMarkerManager`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/theia-marker-manager.ts) propagates received markers to Theia so they show up in the problems view.
Now when the editor is closed, the `TheiaMarkerManager` doesn't remove markers by default, so they stay there.
When the editor is reopened though, the client hasn't requested the markers yet, so there are no markers in the diagram.

In most of the productive applications of GLSP, this mechanism is adapted to my knowledge to the specific needs of the application. Some application apply live validation, thus they request markers on start-up of the editor and send markers from the server after each change. This is the best approach, if validation is rather easy and not a very long running task. If live validation is implemented this way, the problem goes away, because on open, the editor requests markers anyway, so markers would again show up and whenever the client receives markers, the markers are updated in the Theia problems view.
In applications for which validation is very expensive, validation is only performed on request of the client. This is also how it is done for the workflow editor at the moment. I guess, in these scenarios, it would be more consistent to clean up the problem markers on editor clsoe as well.

So I would suggest to open a bug and add the clean-up in the `TheiaMarkerManager` by default, but allow to customize it so it doesn't clean up, which is what applications with live-validation would do.

If we'd want to go one step further in terms of providing generic support for both scenarios in GLSP, we could distinguish API-wise between live validation and on-request validation and only clean-up markers in the `TheiaMarkerManager` that have been contributed by on-request validation.

@tortmayr @martin-fleck-at What do you think?

**Reply by planger** [reply]
Btw, I created https://github.com/eclipse-glsp/glsp/issues/980 to track this enhancement.
